wmo-im/wis2node
# Github Actions Workflow for OGC-API with Observations

### Overview

This Branch and Pull Request were created to address the addition of a Github Actions test workflow that is able to demonstrate that, with our minimal test data, we can go end to end in wis2node.  

Given:

```
  - Test Data (CSVs in this iteration)
    - Metadata (station, discovery)
    - Configurations
```

Workflow (CLI):

```
  - Create discovery metadata
  - Publish discovery metadata as pygeoapi collection
  - Cache station metadata
    - Publish stations metadata as pygeoapi collection
    - Setup data pipeline
    - Transform data (the real workflow will be eventually driven with MQTT, but we need CLI workflow)
      - CSV -> BUFR && GeoJSON
      - Publish the BUFRs (and GeoJSONs) to a web folder for HTTP WAF access and PubSub (MQTT)
      - Publish the GeoJSONs into the pygeoapi backend
```

Successful workflow run should have:

```
  - BUFR file goes in the right directory?
  - GeoJSON file goes in the right directory?
  - Metadata gets published to pygeoapi backend?
  - Observations get published to pygeoapi backend?
  - Errors/edge case checking/handling
```

Thus far, this is what the branch has implemented when the test workflow run locally:

- [x] Create discovery metadata
- [x] Publish discovery metadata as pygeoapi collection
- [x] Cache station metadata
  - [x] Publish stations metadata as pygeoapi collection
  - [x] Setup data pipeline
  - [ ] Transform data (the real workflow will be eventually driven with MQTT, but we need CLI workflow)
    - [x] CSV -> BUFR && GeoJSON
    - [ ] Publish the BUFRs (and GeoJSONs) to a web folder for HTTP WAF access and PubSub (MQTT)
    - [x] Publish the GeoJSONs into the pygeoapi backend

### Where we are now (and issues)

- **Github Action VMs run a docker environment unlike what most users will be using**
  - Although the workflow runs on locally, the vm has trouble with running wis2node and sarracenia ([example workflow with this logged](https://github.com/webb-ben/wis2node/runs/4564215715?check_suite_focus=true)). This means that although the workflow doesn't fail, you will see that the pygeoapi, wis2node, and data-consumer/sarracenia containers are not running or restarting and failing. 
  
![image](https://user-images.githubusercontent.com/40066515/146610719-8addd492-1bcf-4058-9031-d59609de1862.png)

[(run that produced the picture)](https://github.com/wmo-im/wis2node/runs/4564024650?check_suite_focus=true)
  - At what point are we testing against the Github VM and at what point are we testing against the various ways someone might want to deploy wis2node?
  - The VM cannot build docker images from github repos, so I had to use docker images that I built separately for wis2node-ui and sarracenia.
- **What is the test workflow seeking to validate?**
  - This wis2node test workflow is caught between developmental steps of wis2node. Wis2node will be on PYPI, but with the current deployment, the only time wis2node is used, is as a docker container. It seems like we are on track for it to almost always be deployed this way. On the other hand this isn't actually testing against python 3.6, 3.8, or 3.9, just 3.8 which is what is installed by the dockerfile.
  - The `$WIS2NODE...` environment variables are not used in the current `make up && make build` deployment. Should they be mapped into the docker-compose file?
  - Do we need two separate workflows? A first one that builds out the tests without docker to test wis2node against different python versions. Then a second to verify that there is also docker compatibility and a valid end-to-end workflow.

### Specifics about each file (and things that might need to change before this branch can be merged)

- **`tests/load_obs_data.py`:** Reads a file or filesystem and combines all csvs with the like headers into one file, named `observations.csv` inside of the working directory. The headers are that of the first CSV read as well as X, Y, and Z columns retrieved from the station metadata.
- **`tests/Dockerfile`, `tests/build.sh`, and `tests/docker-compose.yml`:** Are used to transform and upload the `observations.csv` into a docker image of `kartoza/postgis`. This could be built out as a separate github action using docker. For minimal viable product, the files are in the tests folder.
- **`docker/pygeoapi/pygeoapi-config.yml`:** Added hardcoded mappinging fro the observations postgis provider. On local machines `localhost` usually needs to be replaced with `host.docker.internal`. Hypothetically you could run pygeoapi with the CSV provider and the output directly from `tests/load_obs_data.py`. That being said, the CSV provider does not have the necessary queryable options to run with wis2node-ui and faces scaling issues.
- **`docker/docker-compose.yml`:** Modifications were only made to run with github actions. Can be reverted once we figure out the other issues I hope.
- **`.github/workflows/tests.yml`:** Added to the workflow are the following steps:
  1. `setup test data ⚙️ PostgreSQL/PostGIS 📦`: Creation of the postgres backend for pygeoapi/observations.
  2. `build Wis2node ⚙️`: Initial build and run of wis2node with `make build` and `make up`.
  3. `generate wis2node collection`: Execution of wis2node scripts on test data. Run to create the station collection, discovery metadata, and observations bufr data.
  4. `run wis2node`: Re run wis2node and print running docker containers.

I am interested to hear others thoughts regarding the best steps going forward.
